src/main/java/org/truffleruby/language/yield/YieldNode.java

```diff
@@ -36,12 +37,27 @@ public final Object executeDispatch(RubyProc block, Object... args) {
 
     @Specialization
     protected Object dispatch(RubyProc block, Object[] argumentsObjects,
-            @Cached CallBlockNode callBlockNode) {
+                              @Cached CallBlockNode callBlockNode,
+                              @Cached(value = "createWarnNodeHelper()", allowUncached = true) WarnNode warnNode) {
+
+        if (getRootNode().getName().equals("<singleton class>")) {
+            if (warnNode.shouldWarnForDeprecation()) {
+                warnNode.warningMessage(
+                        getSourceSection(),
+                        "`yield' in class syntax will not be supported from Ruby 3.0");
+            }
+        }
+
         return callBlockNode.executeCallBlock(
                 block.declarationContext,
                 block,
                 ProcOperations.getSelf(block),
                 null,
                 argumentsObjects);
     }
+
+    protected WarnNode createWarnNodeHelper() {
+        // This shouldn't be needed - try removing it to see the suspected bug in the DSL processor
+        return new WarnNode();
```

Comment:
You would need `@Cached("new()") WarnNode`, because just `@Cached WarnNode` would expect there is `WarnNode.create()` but there isn't right now.

Then there is the more problematic and essential issue, WarnNode doesn't have an uncached version, and it's also not easy to make an uncached version of it. And in fact it's not even a Truffle DSL node, so `@GenerateUncached` wouldn't work.

I think a way to workaround that might be to actually introduce an extra node around the `YieldNode`, in the BodyTranslator that would warn using a WarnNode and then execute the YieldNode, when we detect in the translator that we're inside a singleton class body.
That extra node does not need to support uncached, because it's part of the singleton class body AST, which never needs to be uncached (like other RubyNode).

Also, it avoids the `getRootNode().getName().equals("<singleton class>")` check which seems expensive, and not needed in most cases.

Reply:
Probably simpler, and avoids the need for a new wrapper node: in https://github.com/oracle/truffleruby/blob/d9235b8b2fd7294b7aa215e1bb05f55f31bee60a/src/main/java/org/truffleruby/parser/BodyTranslator.java#L3314 we can see a `YieldExpressionNode` is used to translate the `yield`.
That's a node that is never uncached, so there we can use a WarnNode like in the current approach.

The `getRootNode().getName().equals("<singleton class>")` check could be done only once, and then stored in a `@CompilationFinal Boolean inSingletonClass` on first execution (null -> true|false), or more cleanly but more work, passing a `boolean` to the constructor if we can detect in the translator we are in a singleton class body.

To detect if we are in a singleton class body, there is `TranslatorEnvironment#isModuleBody` but that doesn't tell if it's a singleton class or just some module or class body. We could change that boolean field to be e.g. one of the 3 enum values of OpenModule. I think that would work fine.

There is also the concern of what should happen for `class << self; 1.times { yield }; end`, which could be handled by checking to the first TranslatorEnvironment for which `!TranslatorEnvironment#isBlock()` (i.e., ignoring environment for blocks and looking above). We could have a helper for that, such logic is already used in
https://github.com/oracle/truffleruby/blob/d9235b8b2fd7294b7aa215e1bb05f55f31bee60a/src/main/java/org/truffleruby/parser/BodyTranslator.java#L1953-L1956

Reply:
I hope my comments above are helpful, we can discuss on Slack or have a call to clarify further if you like.

Reply:
Thank you for the comments, @eregon - they were really helpful. I pushed some new commits that has the changes you suggested. 

Here is a quick summary: 
- Added a helper function to `TranslatorEnvironment`
- Removed all changes in `YieldNode` and added the warning logic to `YieldExpressionNode#execute`
- Pass in a new argument `isSingletonClass` into `YieldExpressionNode` from `BodyTranslator` using the helper function

Let me know if the changes aligns to what you were thinking 😅